File: python/logic-2.py
#----------------- Logic-2: --------------------

#make_bricks
#We want to make a row of bricks that is goal inches long. We have a number of small bricks (1 inch each) and big bricks (5
#inches each). Return True if it is possible to make the goal by choosing from the given bricks. This is a little harder than
#it looks and can be done without any loops. See also: Introduction to MakeBricks
def make_bricks(small, big, goal):
  return (5*big)+small>=goal and (goal%5)<= small 


#lone_sum
#Given 3 int values, a b c, return their sum. However, if one of the values is the same as another of the values, it does not
#count towards the sum.
def lone_sum(a, b, c):
  return a*(a!=b and a!=c) + b*(b!=a and b!=c) + c*(c != a and c != b)


#lucky_sum
#Given 3 int values, a b c, return their sum. However, if one of the values is 13 then it does not count towards the sum and
#values to its right do not count. So for example, if b is 13, then both b and c do not count.
def lucky_sum(a, b, c):
  sum = 0
  num = [a,b,c]
  for int in num:
    if int == 13:
      break
    sum += int
  return sum



#no_teen_sum
#Given 3 int values, a b c, return their sum. However, if any of the values is a teen -- in the range 13..19 inclusive -- then
#that value counts as 0, except 15 and 16 do not count as a teens. Write a separate helper "def fix_teen(n):"that takes in an
#int value and returns that value fixed for the teen rule. In this way, you avoid repeating the teen code 3 times (i.e.
#"decomposition"). Define the helper below and at the same indent level as the main no_teen_sum().
def no_teen_sum(a, b, c):
  return fix_teen(a) + fix_teen(b) + fix_teen(c)

def fix_teen(n):
  if (n>=13 and n<=19):
    return n*(n==15 or n==16)
  return n


#round_sum
#For this problem, we'll round an int value up to the next multiple of 10 if its rightmost digit is 5 or more, so 15 rounds up
#to 20. Alternately, round down to the previous multiple of 10 if its rightmost digit is less than 5, so 12 rounds down to 10.
#Given 3 ints, a b c, return the sum of their rounded values. To avoid code repetition, write a separate helper 
#"def round10(num):" and call it 3 times. Write the helper entirely below and at the same indent level as round_sum().
def round_sum(a, b, c):
  return round10(a) + round10(b) + round10(c)

def round10(num):
  mult = num/10
  return 10*(mult+(num-(10*mult)>=5))  #boolean expression adds 1 to mult to round to next 10's



#close_far
#Given three ints, a b c, return True if one of b or c is "close" (differing from a by at most 1), while the other is "far",
#differing from both other values by 2 or more. Note: abs(num) computes the absolute value of a number.
def close_far(a, b, c):
  if compare(a,b,c):
    return True
  b,c = c,b
  return compare(a,b,c)
   
def compare(a,b,c):
    return (abs(a-b)<=1 and abs(a-c)>=2 and abs(c-b)>=2)



#make_chocolate
#We want make a package of goal kilos of chocolate. We have small bars (1 kilo each) and big bars (5 kilos each). Return the
#number of small bars to use, assuming we always use big bars before small bars. Return -1 if it can't be done.
def make_chocolate(small, big, goal):
  if (5*big+small>=goal and (goal%5)<= small):
    if goal/5>big:
      return goal-big*5
    return goal%5
  return -1

File: java/logic-2.java
// Logic-2:


// makeBricks
// We want to make a row of bricks that is goal inches long. We have a number of small bricks (1 inch each) and big bricks (5 inches each). Return true if it is possible to make the goal by choosing from the given bricks. This is a little harder than it looks and can be done without any loops. See also: Introduction to MakeBricks
public boolean makeBricks(int small, int big, int goal) {
  if(goal/5 >= big) return goal-(big*5) <= small;
  else return goal%5 <= small;
}


// loneSum
// Given 3 int values, a b c, return their sum. However, if one of the values is the same as another of the values, it does not count towards the sum.
public int loneSum(int a, int b, int c) {
  int sum = 0;
  if(a!=b && a!=c) sum+=a;
  if(b!=a && b!=c) sum+=b;
  if(c!=a && c!=b) sum+=c;
  return sum;
}


// luckySum
// Given 3 int values, a b c, return their sum. However, if one of the values is 13 then it does not count towards the sum and values to its right do not count. So for example, if b is 13, then both b and c do not count.
public int luckySum(int a, int b, int c) {
  int sum = 0;
  if(a!= 13) sum+=a; 
    else return sum;
  if(b!= 13) sum+=b;
    else return sum;
  if(c!=13) sum+=c;
    else return sum;
  return sum;
}


// noTeenSum
// Given 3 int values, a b c, return their sum. However, if any of the values is a teen -- in the range 13..19 inclusive -- then that value counts as 0, except 15 and 16 do not count as a teens. Write a separate helper "public int fixTeen(int n) {"that takes in an int value and returns that value fixed for the teen rule. In this way, you avoid repeating the teen code 3 times (i.e. "decomposition"). Define the helper below and at the same indent level as the main noTeenSum().
public int noTeenSum(int a, int b, int c) {
  return fixTeen(a)+fixTeen(b)+fixTeen(c);
}

public int fixTeen(int n){
  if(n>=13 && n<=19 && n!=15 && n!=16) return 0;
  return n;
}

// roundSum
// For this problem, we'll round an int value up to the next multiple of 10 if its rightmost digit is 5 or more, so 15 rounds up to 20. Alternately, round down to the previous multiple of 10 if its rightmost digit is less than 5, so 12 rounds down to 10. Given 3 ints, a b c, return the sum of their rounded values. To avoid code repetition, write a separate helper "public int round10(int num) {" and call it 3 times. Write the helper entirely below and at the same indent level as roundSum().
public int roundSum(int a, int b, int c) {
  return round10(a)+round10(b)+round10(c);
}

public int round10(int num){
  if(num%10 >=5) return (num/10 +1)*10; 
  else return (num/10)*10;
}

// closeFar
// Given three ints, a b c, return true if one of b or c is "close" (differing from a by at most 1), while the other is "far", differing from both other values by 2 or more. Note: Math.abs(num) computes the absolute value of a number.
public boolean closeFar(int a, int b, int c) {
  return (Math.abs(a-c)<=1 || Math.abs(a-b)<=1) && (Math.abs(a-c)>=2 || Math.abs(a-b)>=2)
  && Math.abs(b-c)>=2;
}

// blackjack
// Given 2 int values greater than 0, return whichever value is nearest to 21 without going over. Return 0 if they both go over.
public int blackjack(int a, int b) {
  int aVal = (a<=21)?a:0;
  int bVal = (b<=21)?b:0;
  return Math.max(aVal,bVal);
}

// evenlySpaced
// Given three ints, a b c, one of them is small, one is medium and one is large. Return true if the three values are evenly spaced, so the difference between small and medium is the same as the difference between medium and large.
public boolean evenlySpaced(int a, int b, int c) {
  return (a-b) == (b-c) || (a-c)==(c-b) || (b-a)==(a-c) || (b-c)==(c-a) || (c-a)==(a-b) ||
  (c-b)==(b-a);
}


// makeChocolate
// We want make a package of goal kilos of chocolate. We have small bars (1 kilo each) and big bars (5 kilos each). Return the number of small bars to use, assuming we always use big bars before small bars. Return -1 if it can't be done.
public int makeChocolate(int small, int big, int goal) {
  if(goal/5>=big){
    int numSmall = (goal-(big*5));
    return (numSmall <= small) ? numSmall :-1;
  }
  return (goal%5 <=small)? goal%5: -1;
}

